Bounce tracking: Record user activation times.

wanderview opened the pull request:
This information will be used as a signal in other algorithms to avoid deleting storage for these sites.  The deletion algorithms have not been written yet.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/wanderview/nav-tracking-mitigations/pull/35.html" title="Last updated on Apr 4, 2023, 9:02 PM UTC (cf235ab)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/privacycg/nav-tracking-mitigations/35/d53d46a...wanderview:cf235ab.html" title="Last updated on Apr 4, 2023, 9:02 PM UTC (cf235ab)">Diff</a>

jyasskin commented on index.bs:
```diff
@@ -292,16 +295,33 @@ the PrivacyCG.
+1. Assert |document| is not null.
+1. Let |context| be |document|'s [=Document/browsing context=].
+1. If |context| is null, then abort these steps.
+1. Let |topLevelDocument| be |context|'s [=top-level browsing context=]'s
+    [=navigable/active document=].
```
Then this becomes:

```suggestion
1. Let |topLevelDocument| be |context|'s [=navigable/top-level traversable=]'s
    [=navigable/active document=].
```

jyasskin commented on index.bs:
```diff
@@ -292,16 +295,33 @@ the PrivacyCG.
+1. Assert |document| is not null.
+1. Let |context| be |document|'s [=Document/browsing context=].
+1. If |context| is null, then abort these steps.
+1. Let |topLevelDocument| be |context|'s [=top-level browsing context=]'s
+    [=navigable/active document=].
+1. Let |origin| be |topLevelDocument|'s [=Document/origin=].
```
Do you maybe want the |document|'s [=relevant settings object=]'s [[=environment/top level origin=]](https://html.spec.whatwg.org/#concept-environment-top-level-origin)?
It's less clear what that means, but maybe we can define it now.

wanderview replied:
As mentioned separately I actually spec'd the wrong thing here.  I want to avoid recording anything if the activation was not at the top-level.  I'll rework the PR.

wanderview commented on index.bs:
```diff
@@ -292,16 +295,33 @@ the PrivacyCG.
+Append the following steps to the [=activation notification=] steps in the
+[[HTML#user-activation-processing-model|user activation processing model]].
+
+1. Assert |document| is not null.
```
I mainly did this to satisfy the "|document| was only used once warning".  If I make an algorithm then that is solved.

jyasskin replied:
Ah, in general, to silence that explicitly, you can use [`<var ignore>document</var>`](https://speced.github.io/bikeshed/#var-and-algorithms).

wanderview commented:
I've addressed action items and also fixed a few mistakes I made.  In particular I realized we are using schemeless site comparisons because cookies are shared across schemes by default.  @jyasskin PTAL.  Thanks.

jyasskin commented on index.bs:
```diff
@@ -292,16 +295,46 @@ the PrivacyCG.
+
+</div>
+
+Append the following steps to the [=activation notification=] steps in the
```
Use this to avoid needing the new text in the `<pre class="anchors">` section:
```suggestion
Append the following steps to the <a spec="html">activation notification</a> steps in the
```
Documented at https://speced.github.io/bikeshed/#linking-to-unexported-definitions.

jyasskin commented on index.bs:
```diff
@@ -41,6 +41,9 @@ spec: HTTP; urlPrefix: https://httpwg.org/specs/rfc7231.html#
     type: dfn; text: HTTP 3xx statuses; url: status.3xx
 spec: tracking-dnt; urlPrefix: https://www.w3.org/TR/tracking-dnt/#
     type: dfn; text: tracking; url: dfn-tracking
+spec: html; urlPrefix: https://html.spec.whatwg.org/
+    type: dfn
+        text: activation notification; url: #activation-notification
```
Once you're using `<a spec="html">`, you can:
```suggestion
```

jyasskin commented on index.bs:
```diff
@@ -292,16 +295,46 @@ the PrivacyCG.
+To <dfn>record a user activation</dfn> given a [=Document=] |document|, perform
+the following steps:
+
+1. If |document|'s [=node navigable=] is null, then abort these steps.
```
Add a note about why the node-navigable has to be non-null even though you don't use it in the remaining steps.

wanderview replied:
I didn't add the note since I'm now using the navigable.

jyasskin reviewed (approved):
Looks good after 2 small changes. Thanks!

jyasskin commented on index.bs:
```diff
@@ -292,16 +295,46 @@ the PrivacyCG.
+
+1. If |document|'s [=node navigable=] is null, then abort these steps.
+1. Let |settings| be |document|'s [=relevant settings object=].
+1. Let |origin| be |settings|' [=environment/top-level origin=].
```
Oh, I just found https://github.com/whatwg/html/issues/8014, which indicates that we should be moving away from using this, and instead go through the document or window's navigable.
